HPCC-30865 Add generic disk reader for Parquet to hthor

jackdelv opened the pull request:
<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

jackdelv committed:
HPCC-30865 Add generic disk reader for Parquet to hthor

dcamper commented on common/thorhelper/thorread.cpp:
```diff
@@ -1272,6 +1274,205 @@ class AlternativeDiskRowReader : public CInterfaceOf<IDiskRowReader>
+    return this;
+}
+
+const void *ParquetDiskRowReader::nextRow()
```
Minor: Space needed after asterisk (here and in many methods below).

jackdelv replied:
Fixed.

dcamper reviewed (changes requested):
One extremely minor formatting tweak that exists in multiple places.  Otherwise, looks good.

dcamper reviewed (approved):
Looks good.

ghalliday commented on common/thorhelper/CMakeLists.txt:
```diff
@@ -110,6 +110,8 @@ include_directories (
          ${HPCC_SOURCE_DIR}/testing/unittests
          ${HPCC_SOURCE_DIR}/system/security/shared
          ${HPCC_SOURCE_DIR}/fs/dafsclient
+         ${HPCC_SOURCE_DIR}/esp/platform
```
This include does not look right.

ghalliday commented on common/thorhelper/CMakeLists.txt:
```diff
@@ -124,6 +126,7 @@ target_link_libraries ( thorhelper
          roxiemem
          deftype 
          dalibase
+         parquetembed
```
This will need to be conditional on whether parquet is being built.

ghalliday commented on common/thorhelper/thorread.cpp:
```diff
@@ -1272,6 +1274,205 @@ class AlternativeDiskRowReader : public CInterfaceOf<IDiskRowReader>
+
+bool ParquetDiskRowReader::setInputFile(const CLogicalFileSlice & slice, const FieldFilterArray & expectedFilter, unsigned copy)
+{
+    StringBuffer localPath;
```
I'm not sure what this is doing - I suspect it is connected with your testing.

ghalliday commented on common/thorhelper/thorread.cpp:
```diff
@@ -1272,6 +1274,205 @@ class AlternativeDiskRowReader : public CInterfaceOf<IDiskRowReader>
+}
+
+template <class PROCESS>
+const void * ParquetDiskRowReader::inlineNextRow(PROCESS processor)
```
This probably works, I think it will leak at the moment (because next is never freed).

The documentation is inadequate to say the least.

I suspect for parquet you will need 3 different implementations - very similar to the code for csv.  There are 3 cases:
```
const void *nextRow()
```
This would be the same as this function, but returning next instead of calling processor.  It is used by the engines when they are returning rows to the next stage in the processing

```
const void * nextRow(size32_t & resultSize)
```
These are for the case where the row will be thrown away as soon as it is read (e.g., if filtering, or displaying/counting).  This means the output row can be built up in a temporary buffer which is then reused for subsequent calls.

```
const void * nextRow(MemoryBufferBuilder & builder)
```
This is used for dafilesrv - where the resulting rows are being serialised to any output buffer.  In the csv code this could be implemented by one calling the other.  (I can also see a bug in the last of the csv functions that suggests that it has not been used....)

ghalliday reviewed (commented):
Jack, looks like a good start.  A few comments, some of which are probably better to discuss verbally.

ghalliday reviewed (changes requested):
Conclusion from our discussion
* Ensure the parquet references are conditional, so the system will build without it enabled.
* Add a comment to the setInputFile() function to explain what the filename format is and why it is skipping '/'s
* Revisit the nextRow functions looking at the csv functions for a pattern to follow.

jackdelv commented:
@ghalliday I made all the changes that you suggested, but I am having issues getting the Parquet code to link on windows. It is throwing unresolved external symbol errors that suggest the parquetembed.cpp file isn't being properly linked to the thorhelper dll. The only remedy that I can think of is adding ${HPCC_SOURCE_DIR}/plugins/parquet/parquetembed.cpp to the SRCs in thorhelper, but that doesn't seem right.

jackdelv commented:
@ghalliday I fixed the linking issues. Back to you.

ghalliday commented on common/thorhelper/CMakeLists.txt:
```diff
@@ -110,6 +110,10 @@ include_directories (
          ${HPCC_SOURCE_DIR}/testing/unittests
          ${HPCC_SOURCE_DIR}/system/security/shared
          ${HPCC_SOURCE_DIR}/fs/dafsclient
+#ifdef _USE_PARQUET
```
The #ifdef is not correct - since this is a cmake file (it will be treated as a comment).

You need something like:
```
if (USE_PARQUET)
  include_directories ( .... )
endif()
```
See common/dllserver/CMakeList.txt for an example.
Same also goes for line 131 etc.

ghalliday commented on common/thorhelper/CMakeLists.txt:
```diff
@@ -110,6 +110,10 @@ include_directories (
          ${HPCC_SOURCE_DIR}/system/security/shared
          ${HPCC_SOURCE_DIR}/fs/dafsclient
+#ifdef _USE_PARQUET
+         ${HPCC_SOURCE_DIR}/esp/platform
```
Why is the dependency on esp/platform needed?

jackdelv replied:
The esp/platform dependency is needed for tokenserialization.hpp.

ghalliday commented on common/thorhelper/thorread.cpp:
```diff
@@ -14,6 +14,9 @@
     See the License for the specific language governing permissions and
     limitations under the License.
 ############################################################################## */
+#ifdef _USE_PARQUET
```
trivial: Probably clearer, and less likely to have include problems, if it comes after all the other #includes.

jackdelv replied:
I tried moving the #include to the bottom, but that caused include problems in the platform code and apache arrow code. I tried shuffling a few other #includes around in the parquetembed files without any success. How should I approach this in order to find the problem #include statement?

ghalliday reviewed (commented):
@jackdelv looks close.  A few final comments.  Also please do not rebase unless you have to...

@GordonSmith please can you double check the cmake integration is what you would expect.

GordonSmith commented on common/thorhelper/CMakeLists.txt:
```diff
@@ -126,6 +133,10 @@ target_link_libraries ( thorhelper
 
+if (USE_PARQUET)
+    target_link_libraries( thorhelper parquetembed )
+endif()
```
@ghalliday - On Linux will this cause issues if the parquetembed.so is dynamically loaded/unloaded by a WU?

GordonSmith reviewed (approved):
CMake files looks fine - I had one question about dynamic loading though...

ghalliday reviewed (commented):
I don't mind if you move the #include.  I think it is worth removing the dependency on  tokenserialization.hpp

jackdelv commented:
@ghalliday I removed the dependency on esp, and moved the #include for parquetembed.hpp. Should I open a JIRA for removing the esp dependency from mongodb/couchbase as well? Back to you.

AttilaVamos commented:
I can see some build problem in the Guerrilla Smoketest related this PR:
'''
[ 44%] Building CXX object common/thorhelper/CMakeFiles/thorhelper.dir/thorstats.cpp.o
In file included from /home/centos/smoketest/PR-18233/build/vcpkg_installed/x64-linux-dynamic/include/arrow/util/async_generator.h:32,
                 from /home/centos/smoketest/PR-18233/build/vcpkg_installed/x64-linux-dynamic/include/arrow/acero/options.h:34,
                 from /home/centos/smoketest/PR-18233/build/vcpkg_installed/x64-linux-dynamic/include/arrow/dataset/scanner.h:28,
                 from /home/centos/smoketest/PR-18233/build/vcpkg_installed/x64-linux-dynamic/include/arrow/dataset/file_base.h:31,
                 from /home/centos/smoketest/PR-18233/build/vcpkg_installed/x64-linux-dynamic/include/arrow/dataset/api.h:25,
                 from /home/centos/smoketest/PR-18233/HPCC-Platform/plugins/parquet/parquetembed.hpp:26,
                 from /home/centos/smoketest/PR-18233/HPCC-Platform/common/thorhelper/thorread.cpp:39:
/home/centos/smoketest/PR-18233/build/vcpkg_installed/x64-linux-dynamic/include/arrow/util/io_util.h:122:87: error: macro "DeleteFile" passed 2 arguments, but takes just 1
  122 | Result<bool> DeleteFile(const PlatformFilename& file_path, bool allow_not_found = true);
      |                                                                                       ^
In file included from /home/centos/smoketest/PR-18233/HPCC-Platform/system/jlib/jliball.hpp:22,
                 from /home/centos/smoketest/PR-18233/HPCC-Platform/common/thorhelper/thorread.cpp:17:
/home/centos/smoketest/PR-18233/HPCC-Platform/system/include/platform.h:349: note: macro "DeleteFile" defined here
  349 | #define DeleteFile(name)    (unlink(name)==0)
      | 
'''
Maybe a rebase issue? Some related changes in platform.h merged yesterday.
@ghalliday any thoughts?

ghalliday reviewed (approved):
@jackdelv Thanks.  That all looks good.  Please can you squash and I will scan again and then merge.

jackdelv commented:
@ghalliday Squashed.

AttilaVamos commented:
Unfortunately it is still failing to build on CentOS:
```
	/home/centos/smoketest/PR-18233/build/vcpkg_installed/x64-linux-dynamic/include/arrow/util/io_util.h:122:87: error: macro "DeleteFile" passed 2 arguments, but takes just 1
	/home/centos/smoketest/PR-18233/HPCC-Platform/system/include/platform.h:364:17: error: expected identifier before â€˜boolâ€™
	/home/centos/smoketest/PR-18233/HPCC-Platform/system/include/platform.h:364:17: error: expected â€˜}â€™ before â€˜boolâ€™
	/home/centos/smoketest/PR-18233/build/vcpkg_installed/x64-linux-dynamic/include/arrow/type_fwd.h:293:9: error: expected unqualified-id before â€˜,â€™ token
	/home/centos/smoketest/PR-18233/build/vcpkg_installed/x64-linux-dynamic/include/arrow/type_fwd.h:422:26: error: â€˜Typeâ€™ was not declared in this scope; did you mean â€˜arrow::Typeâ€™?
	/home/centos/smoketest/PR-18233/build/vcpkg_installed/x64-linux-dynamic/include/arrow/type_fwd.h:422:36: error: template argument 1 is invalid
	/home/centos/smoketest/PR-18233/build/vcpkg_installed/x64-linux-dynamic/include/arrow/type_fwd.h:422:36: error: template argument 2 is invalid
	/home/centos/smoketest/PR-18233/build/vcpkg_installed/x64-linux-dynamic/include/arrow/type_fwd.h:430:36: error: â€˜DataTypeâ€™ was not declared in this scope; did you mean â€˜arrow::DataTypeâ€™?
	/home/centos/smoketest/PR-18233/build/vcpkg_installed/x64-linux-dynamic/include/arrow/type_fwd.h:430:44: error: template argument 1 is invalid
	/home/centos/smoketest/PR-18233/build/vcpkg_installed/x64-linux-dynamic/include/arrow/type_fwd.h:432:36: error: â€˜DataTypeâ€™ was not declared in this scope; did you mean â€˜arrow::DataTypeâ€™?
```
GCC:  gcc (GCC) 11.2.1 20220127 (Red Hat 11.2.1-9)

ghalliday commented:
@jackdelv where are you seeing that error from?  I think it is not picking up the most recent sources.  I don't see a failure in the github actions.

jackdelv commented:
@ghalliday After I applied your changes I stopped getting the error. However, I didn’t rebase to apply it, so it’s possible that could fix it.
